_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,457 @@
+
+[^stopping]:Hyperband's theory answers "how many models should be stopped?" and "when should they be stopped?"
+
+[^follow-up]:See Figure 1 of [the BOHB paper][BOHB paper] and [a paper][blippar] from an augmented reality company.
+
+[^hyperband-figs]:See Figures 4, 7 and 8 in "[Hyperband: A Novel Bandit-Based Approach toHyperparameter Optimization][hyperband-paper]."
+
+[blippar]:https://arxiv.org/pdf/1801.01596.pdf
+[BOHB paper]:http://proceedings.mlr.press/v80/falkner18a/falkner18a.pdf
+[nni]:https://nni.readthedocs.io/en/latest/Tuner/HyperbandAdvisor.html
+
+<img width="80%" src="/images/2020-model-selection/hyperband-fig-7-8.png"
+ style="display: block; margin-left: auto; margin-right: auto;" />
+
+<div style="max-width: 80%; word-wrap: break-word;" style="text-align: center;">
+<sup>"hyperband (finite)" is similar Dask-ML's default implementation, and
+"bracket s=4" is similar to Ray's default implementation. SMAC,[^smac]
+spearmint,[^spearmint] and TPE[^tpe] are popular Bayesian algorithms.
+"random" is random search, and "random 2x" does twice as much
+computation.</sup>
+</div>
+
+[^smac]:SMAC is described in "[Sequential Model-Based Optimization forGeneral Algorithm Configuration][SMAC-paper]," and is available [in AutoML][SMAC-automl].
+[^spearmint]:Spearmint is described in "[Practical Bayesian Optimization of MachineLearning Algorithms][pbo]," and is available in [HIPS/spearmint].
+[^tpe]:TPE is described in Section 4 of "[Algorithms for Hyperparameter Optimization][aho]," and is available [through Hyperopt][hyperopt].
+[SMAC-paper]:https://www.cs.ubc.ca/~hutter/papers/10-TR-SMAC.pdf
+[SMAC-automl]:https://www.automl.org/automated-algorithm-design/algorithm-configuration/smac/
+[spearmint]:https://github.com/HIPS/Spearmint
+[aho]:http://papers.nips.cc/paper/4443-algorithms-for-hyper-parameter-optimization.pdf
+[pbo]:https://papers.nips.cc/paper/4522-practical-bayesian-optimization-of-machine-learning-algorithms.pdf
+[HIPS/spearmint]:https://github.com/HIPS/Spearmint
+[hyperopt]:http://hyperopt.github.io/hyperopt/
+
+The second method of avoiding unnecessary work is called "Bayesian
+optimization." These are more difficult in any distributed system because of
+the serial nature of Bayesian sampling algorithms.  Bayesian sampling treats
+models as black boxes, and tries to deduce the set of inputs that maximize the
+black box's quality. This is difficult in any distributed system because if
+there were an infinite number of workers, the Bayesian sampling algorithm
+won't have time to deduce anything after initializing models.
+
+Notably, Bayesian optimization algorithms have been adapted the Hyperband
+algorithm. For example, AutoML's [BOHB] uses TPE and Hyperband in tandem.  Even
+better, they have an intelligent way of scaling to distributed systems given
+Hyperband's parallel architecture[^scaling]<sup>,</sup>[^ray-bayesian] -- it's
+future to integrate this Bayesian sampling into Dask-ML.
```

Comment:
missing a word here. "future work" I think.

Reply:
I've removed this whole paragraph on improvements to Bayesian sampling. As you correctly pointed out, the goal is to correct the false statements, not detail improvements or other algorithms.